File: pine-java/src/main/java/page/liam/pine/operators/MergeDedup.java
package page.liam.pine.operators;

import page.liam.pine.AbstractOperator;
import page.liam.pine.CancellationToken;
import page.liam.pine.OperatorInput;
import page.liam.pine.OperatorOutput;
import page.liam.pine.OperatorParams;

import java.util.*;

/**
 * Operator: merge_dedup
 * Metadata contract
 *   CommonInput:  []
 *   CommonOutput: []
 *   ItemInput:    [item_id, _source]
 *   ItemOutput:   [item_id]
 */
public class MergeDedup extends AbstractOperator implements page.liam.pine.ConsumesRowSet, page.liam.pine.MutatesRowSet {
    private String strategy;

    @Override
    public void init(OperatorParams params) {
        strategy = params.getString("strategy", "first");
        if (!"first".equals(strategy)) {
            throw new IllegalArgumentException("merge_dedup: unsupported strategy \"" + strategy + "\"");
        }
    }

    @Override
    public void execute(CancellationToken token, OperatorInput input, OperatorOutput output) {
        String dedupBy = itemInput().get(0);
        Set<Object> seen = new LinkedHashSet<>();
        for (int i = 0; i < input.itemCount(); i++) {
            Object key = normalizeKey(input.item(i, dedupBy));
            if (seen.contains(key)) {
                output.removeItem(i);
            } else {
                seen.add(key);
            }
        }
    }

    private static Object normalizeKey(Object v) {
        if (v instanceof Number) {
            double d = ((Number) v).doubleValue();
            // IEEE 754 -0.0 and +0.0 are semantically equal but
            // Double.hashCode(-0.0) != Double.hashCode(0.0), so HashSet
            // treats them as different keys. Go's map uses == for float
            // keys which considers -0 == +0. Normalize to +0.0.
            if (d == 0.0) d = 0.0;
            return d;
        }
        return v;
    }
}

File: pine-cpp/operators/merge/merge_dedup.cpp
#include "pine/operator.hpp"

#include <unordered_set>

#include "operators/_helpers.hpp"

namespace pine {

class MergeDedupOp : public Operator, public ConsumesRowSet, public MutatesRowSet {
 public:
  void init(const OperatorConfig& cfg) override {
    op_name_ = cfg.name;
    field_ = cfg.metadata.item_input.at(0);
  }
  void execute(const OperatorInput& input, OperatorOutput& out) override {
    const std::size_t n = input.item_count();
    std::unordered_set<std::string> seen;
    seen.reserve(n);
    // Batched column access: one lock + one lookup instead of per-element
    // item() calls.
    std::vector<Variant> col = input.item_column(field_);
    for (std::size_t i = 0; i < n; ++i) {
      auto [_it, inserted] = seen.emplace(operators::dedup_key(col[i]));
      if (!inserted) {
        out.remove_item(static_cast<int>(i));
      }
    }
  }

 private:
  std::string op_name_;
  std::string field_;
};

static const OperatorSchema k_merge_dedup_schema{
    .name = "merge_dedup",
    .type = OpType::Merge,
    .description = "Deduplicates items by a key field, keeping the first occurrence.",
    .params =
        {
            {"strategy",
             {.type = "string",
              .required = false,
              .default_value = Variant("first"),
              .description = "Dedup strategy \xe2\x80\x94 \"first\" keeps first occurrence."}},
        },
    .metadata = {.common_input = "[]",
                 .common_output = "[]",
                 .item_input = "[item_id, _source]",
                 .item_output = "[item_id]"},
};
PINE_REGISTER_OPERATOR_T(MergeDedupOp, k_merge_dedup_schema)

}  // namespace pine
